File: electron/tests/mods_test.ts
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// The mods bridge — chiefly its CONTAINMENT check.
//
// One value crosses from the page INWARD in this whole feature: the `folder` a
// PUBLISH names. Everything else travels outward. So that is the one string
// that turns page-supplied text into a filesystem operation — an upload to a
// public Workshop, no less — and it gets a real test rather than a careful
// comment, exactly like `webroot.ts`'s path check.
//
// `electron`'s `app` is stubbed: these tests run under plain vitest with no
// Electron runtime, which is also the arrangement that keeps the desktop
// check job cheap (no 100 MB binary download in CI).

import {
  cpSync,
  existsSync,
  mkdirSync,
  mkdtempSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";

import {
  createModsBridge,
  localModsDir,
  portableModsDir,
  portableModsPath,
  type ModsEvent,
} from "../src/mods";

/** The suite runs UNPACKAGED, where every platform has a portable folder — so
 * this narrowing is a fact of the fixture, not an assumption about the app. */
const portableDir = (): string => portableModsDir() as string;
import { zip, zipDir } from "./zip-fixture";

/** The worked mod in the repo, copied in as the player's own local mod.
 * `__dirname`-relative because this suite compiles to CommonJS. */
const EXAMPLE = join(__dirname, "..", "..", "mod", "examples", "greenhouse");

type InstalledModShape = {
  key: string;
  folder: string;
  source: string;
  bundle: unknown;
  errors: string[];
};

let userData: string;
/** `portableDir()` is cwd-relative when unpackaged, so the suite runs from
 * a directory of its own — that IS the "beside the game" folder here. */
let portableHome: string;
const originalCwd = process.cwd();

const opened: string[] = [];
vi.mock("electron", () => ({
  app: {
    getPath: () => userData,
    isPackaged: false,
  },
  shell: {
    openPath: (dir: string) => {
      opened.push(dir);
      return Promise.resolve("");
    },
    openExternal: () => Promise.resolve(),
  },
}));

const temps: string[] = [];

beforeAll(() => {
  userData = mkdtempSync(join(tmpdir(), "gis-userdata-"));
  portableHome = mkdtempSync(join(tmpdir(), "gis-install-"));
  temps.push(userData, portableHome);
  process.chdir(portableHome);
});

afterAll(() => {
  process.chdir(originalCwd);
  for (const dir of temps) rmSync(dir, { recursive: true, force: true });
});

describe("the local mods folder", () => {
  it("is created on first look, so the path in the docs exists to be opened", async () => {
    const dir = localModsDir();
    expect(dir).toBe(join(userData, "mods"));
    // Calling it again must not throw on the now-existing directory.
    expect(localModsDir()).toBe(dir);
  });
});

describe("where the folder beside the game is", () => {
  const packaged = (platform: NodeJS.Platform, exe: string) =>
    portableModsPath({ packaged: true, platform, exe, cwd: "/repo" });

  it("is the install folder on Windows and Linux", () => {
    expect(packaged("win32", "C:\\Games\\adastrail\\adastrail.exe")).toBe(
      "C:\\Games\\adastrail\\mods",
    );
    expect(packaged("linux", "/opt/adastrail/adastrail")).toBe(
      "/opt/adastrail/mods",
    );
  });

  it("does NOT exist on macOS, where the app sits in /Applications", () => {
    // Beside the app is a system folder the player does not own, and inside
    // the bundle would break the signature it is notarized under. Application
    // Support is the whole answer there.
    expect(
      packaged(
        "darwin",
        "/Applications/Adas Trail.app/Contents/MacOS/adastrail",
      ),
    ).toBeNull();
  });

  it("is the working directory when unpackaged, on every platform", () => {
    // A checkout is a developer's own tree, not an installed app — so even
    // macOS has one here.
    for (const platform of ["darwin", "linux"] as NodeJS.Platform[]) {
      expect(
        portableModsPath({ packaged: false, platform, exe: "", cwd: "/repo" }),
      ).toBe("/repo/mods");
    }
    expect(
      portableModsPath({
        packaged: false,
        platform: "win32",
        exe: "",
        cwd: "C:\\repo",
      }),
    ).toBe("C:\\repo\\mods");
  });
});

describe("listing", () => {
  const list = (requestId: number) =>
    new Promise<{ mods: InstalledModShape[] }>((resolve) => {
      const bridge = createModsBridge((event: ModsEvent) =>
        resolve(event as unknown as { mods: InstalledModShape[] }),
      );
      bridge.handle({ action: "list", requestId });
    });

  it("compiles a real mod through the real compiler", async () => {
    // The end-to-end shape of the thing: the shell dynamically imports the ESM
    // compiler from outside its own tree, reads the reference catalog, and
    // hands the page compiled JSON. If the paths in resources.ts are wrong, or
    // the .mjs cannot be imported from this CommonJS main process, this is
    // where it shows — not on a player's machine.
    cpSync(EXAMPLE, join(localModsDir(), "greenhouse"), { recursive: true });

    const mod = (await list(1)).mods.find((m) => m.key === "local:greenhouse");
    expect(mod).toBeDefined();
    expect(mod!.errors).toEqual([]);
    expect(mod!.source).toBe("local");
    expect((mod!.bundle as { id: string }).id).toBe("greenhouse");
  });

  it("keeps a mod that did not compile, with its errors", async () => {
    // A subscription that fails to build must still reach the MODS screen —
    // an empty list leaves the player with no way to learn why.
    const broken = join(localModsDir(), "broken");
    mkdirSync(broken, { recursive: true });
    writeFileSync(join(broken, "mod.yaml"), "id: broken\nname: BROKEN\n");

    const mod = (await list(2)).mods.find((m) => m.key === "local:broken");
    expect(mod).toBeDefined();
    expect(mod!.bundle).toBeNull();
    expect(mod!.errors.length).toBeGreaterThan(0);
  });

  it("reads a mod that arrived as a .zip, wrapper folder and all", async () => {
    // What a friend actually sends: the mod folder compressed, so everything
    // sits under one top-level directory inside the archive.
    mkdirSync(portableDir(), { recursive: true });
    writeFileSync(
      join(portableDir(), "greenhouse.zip"),
      zipDir(EXAMPLE, "greenhouse"),
    );

    const mod = (await list(4)).mods.find(
      (m) => m.key === "portable:greenhouse.zip",
    );
    expect(mod).toBeDefined();
    expect(mod!.errors).toEqual([]);
    expect((mod!.bundle as { id: string }).id).toBe("greenhouse");
    // Unpacked into the archive cache, NOT into the authoring folder — which
    // is what keeps a received mod out of the publish path.
    expect(mod!.folder.startsWith(join(userData, "mod-archives"))).toBe(true);
  });

  it("calls a .zip portable even in the authoring folder — it is not publishable", async () => {
    writeFileSync(
      join(localModsDir(), "sent-to-me.zip"),
      zipDir(EXAMPLE, "sent-to-me"),
    );
    const mod = (await list(8)).mods.find(
      (m) => m.key === "local:sent-to-me.zip",
    );
    expect(mod).toBeDefined();
    expect(mod!.source).toBe("portable");
    expect(mod!.errors).toEqual([]);
  });

  it("finds a plain folder beside the game, and calls it portable", async () => {
    cpSync(EXAMPLE, join(portableDir(), "beside"), { recursive: true });
    const mod = (await list(5)).mods.find((m) => m.key === "portable:beside");
    expect(mod).toBeDefined();
    expect(mod!.source).toBe("portable");
    expect(mod!.errors).toEqual([]);
  });

  it("reports a .zip that is not a mod, rather than ignoring the file", async () => {
    // Unlike a nameless directory: a file put in the mods folder was put there
    // to be played, so silence would be the wrong answer.
    writeFileSync(
      join(portableDir(), "holiday-photos.zip"),
      zip([{ name: "beach.jpg", body: Buffer.from("not a mod") }]),
    );
    const mod = (await list(6)).mods.find(
      (m) => m.key === "portable:holiday-photos.zip",
    );
    expect(mod).toBeDefined();
    expect(mod!.bundle).toBeNull();
    expect(mod!.errors[0]).toMatch(/no mod.yaml/);
  });

  it("ignores a directory that is not a mod at all", async () => {
    mkdirSync(join(localModsDir(), "just-some-notes"), { recursive: true });
    const mods = (await list(3)).mods;
    expect(mods.some((m) => m.key === "local:just-some-notes")).toBe(false);
  });
});

describe("showing a folder in the file manager", () => {
  const reveal = (which: "local" | "portable") => {
    const bridge = createModsBridge(() => {});
    bridge.handle({ action: "reveal", requestId: 0, which });
  };

  it("opens the folder the page NAMED, resolved here", () => {
    opened.length = 0;
    reveal("local");
    expect(opened).toEqual([localModsDir()]);
    reveal("portable");
    expect(opened[1]).toBe(portableDir());
  });

  it("creates the folder first, so it is never opened into nothing", () => {
    rmSync(portableDir(), { recursive: true, force: true });
    opened.length = 0;
    reveal("portable");
    expect(existsSync(portableDir())).toBe(true);
  });

  it("takes no path from the page — only the two names", () => {
    // The request carries `which`, never a directory: there is no field here
    // that could name /etc or somebody's documents.
    opened.length = 0;
    const bridge = createModsBridge(() => {});
    bridge.handle({
      action: "reveal",
      requestId: 0,
      folder: "/etc",
    } as Parameters<typeof bridge.handle>[0]);
    // With no `which`, it falls back to the authoring folder — never `folder`.
    expect(opened).toEqual([localModsDir()]);
  });
});

describe("publish refuses a folder outside the player's mods directory", () => {
  /** Drive one publish request through the bridge and read the reply. */
  const publish = (folder: string) =>
    new Promise<Record<string, unknown>>((resolve) => {
      const bridge = createModsBridge((event: ModsEvent) =>
        resolve(event as unknown as Record<string, unknown>),
      );
      bridge.handle({ action: "publish", requestId: 1, folder });
    });

  it("a sibling directory whose name merely STARTS with the mods path", async () => {
    // `…/mods-elsewhere` shares a string prefix with `…/mods` and would pass a
    // naive `startsWith`. It must not.
    const sneaky = join(userData, "mods-elsewhere");
    mkdirSync(sneaky, { recursive: true });
    writeFileSync(join(sneaky, "mod.yaml"), "id: sneaky\n");
    const reply = await publish(sneaky);
    expect(reply.ok).toBe(false);
    expect(reply.reason).toBe("not-a-mod");
  });

  it("a traversal back out of the mods directory", async () => {
    const escape = join(userData, "mods", "..", "..");
    const reply = await publish(escape);
    expect(reply.ok).toBe(false);
    expect(reply.reason).toBe("not-a-mod");
  });

  it("the cache a received .zip was unpacked into", async () => {
    // A mod that ARRIVED is not a mod the player authored. The extraction
    // cache sits beside `mods/` rather than inside it precisely so the
    // containment check keeps it out of the Workshop.
    mkdirSync(portableDir(), { recursive: true });
    writeFileSync(join(portableDir(), "sendme.zip"), zipDir(EXAMPLE, "sendme"));
    const listed = await new Promise<{ mods: InstalledModShape[] }>(
      (resolve) => {
        const bridge = createModsBridge((event: ModsEvent) =>
          resolve(event as unknown as { mods: InstalledModShape[] }),
        );
        bridge.handle({ action: "list", requestId: 7 });
      },
    );
    const unpacked = listed.mods.find((m) => m.key === "portable:sendme.zip");
    expect(unpacked?.folder).toBeDefined();

    const reply = await publish(unpacked!.folder);
    expect(reply.ok).toBe(false);
    expect(reply.reason).toBe("not-a-mod");
  });

  it("a folder beside the game rather than in the authoring directory", async () => {
    cpSync(EXAMPLE, join(portableDir(), "received"), { recursive: true });
    const reply = await publish(join(portableDir(), "received"));
    expect(reply.ok).toBe(false);
    expect(reply.reason).toBe("not-a-mod");
  });

  it("an empty folder", async () => {
    const reply = await publish("");
    expect(reply.ok).toBe(false);
    expect(reply.reason).toBe("not-a-mod");
  });

  it("a path inside the mods directory that does not exist", async () => {
    const reply = await publish(join(userData, "mods", "no-such-mod"));
    expect(reply.ok).toBe(false);
    expect(reply.reason).toBe("not-a-mod");
  });
});

File: tauri/shell/tests/mods_test.rs
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
//! THE MODS SEAM's rules — which folders hold mods, which of them may be
//! published, and what crosses to the page.
//!
//! The one that matters most is [`is_local_mod`]: `folder` is the ONE path the
//! page hands INWARD, publishing is an upload, and a folder outside the
//! player's own mods directory is not something the page has any business
//! naming.

use std::path::{Path, PathBuf};

use adastrail_shell::mods::{
    archive_stamp, folder_key, is_local_mod, list_event, local_mods_dir, parse, portable_mods_path,
    publish_event, read_item_id, safe_slug, workshop_url, InstalledMod, ModSource, PortableEnv,
    PublishOutcome, RevealTarget,
};
use serde_json::{json, Value};

fn env(packaged: bool, platform: &str) -> PortableEnv {
    PortableEnv {
        packaged,
        platform: platform.to_string(),
        exe_dir: PathBuf::from("/Applications/Adas Trail.app/Contents/MacOS"),
        cwd: PathBuf::from("/home/nic/game"),
    }
}

#[test]
fn a_reveal_names_one_of_two_folders_and_never_a_path() {
    // The page picking from a closed set is what keeps "open this in the file
    // manager" from becoming "open anything on this disk in the file manager".
    let local = parse(&json!({ "action": "reveal", "which": "local" }));
    assert_eq!(local.which, RevealTarget::Local);
    assert_eq!(
        parse(&json!({ "action": "reveal", "which": "portable" })).which,
        RevealTarget::Portable
    );
    // Anything else — including a path — falls back to the authoring folder
    // rather than being honoured.
    assert_eq!(
        parse(&json!({ "action": "reveal", "which": "/etc" })).which,
        RevealTarget::Local
    );
    assert_eq!(
        parse(&json!({ "action": "reveal" })).which,
        RevealTarget::Local
    );
}

#[test]
fn a_publish_may_only_name_a_folder_inside_the_authoring_directory() {
    let root = Path::new("/home/nic/.local/share/adastrail-tauri/mods");
    assert!(is_local_mod(
        Path::new("/home/nic/.local/share/adastrail-tauri/mods/greenhouse"),
        root
    ));

    // The directory itself is not a mod, and neither is anything beside it.
    assert!(!is_local_mod(root, root));
    assert!(!is_local_mod(
        Path::new("/home/nic/.local/share/adastrail-tauri/mods-elsewhere/x"),
        root
    ));
    assert!(!is_local_mod(Path::new("/etc"), root));
    // …and `..` cannot climb out of it, which is the whole reason the check is
    // lexical rather than a string prefix.
    assert!(!is_local_mod(
        Path::new("/home/nic/.local/share/adastrail-tauri/mods/../../secrets"),
        root
    ));
    // A path that leans on `.` and `..` and still lands inside is fine.
    assert!(is_local_mod(
        Path::new("/home/nic/.local/share/adastrail-tauri/mods/./a/../greenhouse"),
        root
    ));
}

#[test]
fn macos_has_no_folder_beside_the_app_and_that_is_the_platform_rather_than_an_omission() {
    // An installed app lives in /Applications, which the player does not own —
    // and writing inside the bundle breaks the signature it is notarized under.
    assert_eq!(portable_mods_path(&env(true, "macos")), None);
    assert_eq!(
        portable_mods_path(&env(true, "windows")),
        Some(PathBuf::from(
            "/Applications/Adas Trail.app/Contents/MacOS/mods"
        ))
    );
    // Unpackaged is a developer's own tree on every platform, macOS included.
    assert_eq!(
        portable_mods_path(&env(false, "macos")),
        Some(PathBuf::from("/home/nic/game/mods"))
    );
}

#[test]
fn the_authoring_folder_hangs_off_the_apps_own_user_data() {
    assert_eq!(
        local_mods_dir(Path::new("/home/nic/.local/share/adastrail-tauri")),
        PathBuf::from("/home/nic/.local/share/adastrail-tauri/mods")
    );
}

#[test]
fn a_mod_that_did_not_compile_crosses_with_its_reasons() {
    // A silent omission would leave a subscriber with an empty list and no way
    // to find out why the thing they paid attention to is not playable.
    let broken = InstalledMod {
        key: "3141592".to_string(),
        folder: "/steam/workshop/3141592".to_string(),
        source: ModSource::Workshop,
        bundle: Value::Null,
        errors: vec!["items/spade.yaml: unknown field `dmg`".to_string()],
        needs_update: true,
    };
    let event = list_event(
        5,
        std::slice::from_ref(&broken),
        Path::new("/home/nic/mods"),
        None,
    );
    assert_eq!(event["ok"], json!(true), "the LIST succeeded");
    assert_eq!(event["mods"][0]["bundle"], Value::Null);
    assert_eq!(
        event["mods"][0]["errors"][0],
        json!("items/spade.yaml: unknown field `dmg`")
    );
    assert_eq!(event["mods"][0]["source"], json!("workshop"));
    assert_eq!(event["mods"][0]["needsUpdate"], json!(true));
    // macOS has no portable folder, and the page draws one row fewer for it.
    assert_eq!(event["folders"]["portable"], Value::Null);
    assert_eq!(event["folders"]["local"], json!("/home/nic/mods"));
}

#[test]
fn the_agreement_is_its_own_outcome_because_it_is_the_one_the_player_must_act_on() {
    // Steam refuses to SHOW an item until its author accepts the Workshop terms
    // in a browser — and the item exists in the meantime, invisible.
    let event = publish_event(
        6,
        &PublishOutcome::Published {
            item_id: "3141592".to_string(),
            needs_to_accept_agreement: true,
        },
    );
    assert_eq!(event["ok"], json!(true));
    assert_eq!(event["itemId"], json!("3141592"));
    assert_eq!(event["needsToAcceptAgreement"], json!(true));

    let refused = publish_event(
        6,
        &PublishOutcome::Refused {
            reason: "not-a-mod",
            detail: None,
        },
    );
    assert_eq!(refused["reason"], json!("not-a-mod"));
    assert!(refused.get("detail").is_none());
}

#[test]
fn the_workshop_door_is_built_from_our_own_app_id() {
    // The one thing this action must not become is an open-arbitrary-URL
    // channel, so nothing the page sent reaches the string.
    assert_eq!(workshop_url(480), "steam://url/SteamWorkshopPage/480");
}

#[test]
fn an_archive_is_re_extracted_when_the_file_changes_and_not_before() {
    // Keyed by size and modification time, so replacing a zip with a newer one
    // is picked up on the next list and a launch that changed nothing pays a
    // stat instead of an unpack.
    assert_eq!(
        archive_stamp(4_096, 1_700_000_000_000),
        "4096-1700000000000"
    );
    assert_ne!(archive_stamp(4_096, 1), archive_stamp(4_097, 1));
}

#[test]
fn an_archives_own_name_becomes_a_name_and_nothing_else() {
    // It is a filename the player chose, and it is about to be a path.
    assert_eq!(safe_slug("greenhouse"), "greenhouse");
    assert_eq!(safe_slug("green house v2"), "green-house-v2");
    assert_eq!(safe_slug("../../etc/passwd"), "etc-passwd");
    assert_eq!(safe_slug("...."), "archive");
    assert_eq!(safe_slug(""), "archive");
    assert!(safe_slug(&"x".repeat(200)).len() <= 64);
}

#[test]
fn a_row_is_keyed_by_where_it_came_from() {
    assert_eq!(
        folder_key(ModSource::Local, "greenhouse"),
        "local:greenhouse"
    );
    assert_eq!(
        folder_key(ModSource::Portable, "sent.zip"),
        "portable:sent.zip"
    );
}

#[test]
fn a_remembered_workshop_id_is_a_number_or_it_is_nothing() {
    // The id lives beside the mod so that copying the folder to another machine
    // still updates the same item rather than minting a second one — and a
    // file somebody edited must not become a publish against a stranger's item.
    assert_eq!(read_item_id("3141592\n"), Some("3141592".to_string()));
    assert_eq!(read_item_id("  3141592  "), Some("3141592".to_string()));
    assert_eq!(read_item_id(""), None);
    assert_eq!(read_item_id("not an id"), None);
    assert_eq!(read_item_id("314 592"), None);
}
